### PR TITLE
MA: Cataclysm in one state

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

Mods "MA: The Absolute State of DDA: Part 1"

#### Purpose of change

Overmap overhaul mod for Cataclysm: DDA that uses some features of Massachusetts state when generating overmaps.

#### Describe the solution

See commit messages.

#### Testing

1. Start game in new world with MA mod - during new game start you can use `C` to select city you want to spawn in (start location can have some restrictions attached - city size, city distance, allowed z-levels).
2. Hop to specific overmaps with cities (use attached spreadsheet for reference) using new debug command for city teleport and make sure cities are there as expected.

#### Additional context

It is just a first step. Terrain goes second.

#### Screenshots

image

image

image

```JSON
{
  {
    "type": "city",
    "id": "MA_35_BOSTON",
    "database_id": 35,
    "name": "BOSTON",
    "population": 617594,
    "size": 16,
    "pos_om": [ 46, 15 ],
    "pos": [ 57, 94 ]
  }
}
```

### CODE REVIEW

**Comment by Zireael07** [comment]
Whoa, this might get me back to actually, y'know, playing the game instead of just spinning up "test" characters/scenarios. :fireworks:

File: src/game.cpp [hunk 1]
```diff
@@ -697,7 +697,36 @@ bool game::start_game()
...
+        uilist select_city;
+        select_city.desc_enabled = true;
+        select_city.allow_cancel = false;
+        select_city.title = _( "Select a starting city" );
+        for( const auto &c : sorted_cities ) {
+            uilist_entry entry( c.id, true, -1, c.name,
+                                string_format(
+                                    _( "Location: <color_white>%s</color>:<color_white>%s</color>" ),
+                                    c.pos_om.to_string(), c.pos.to_string(), c.size ),
+                                string_format( _( "(pop. <color_white>%s</color>)" ), c.population ) );
```
Comment (actual-nh) [inline src/game.cpp:720]:
```suggestion
                                string_format( _( "(pop <color_white>%s</color>)" ), c.population ) );
```

**Review by actual-nh — COMMENTED** [review]
Interesting!

**Comment by kevingranade** [comment]
You missed out, could have been "MA: The Absolute State of DDA"

File: src/game.cpp [hunk 2]
```diff
@@ -697,7 +697,40 @@ bool game::start_game()
...
+    point_abs_om world_origin;
+    overmap &omap = overmap_buffer.get( world_origin );
+    std::vector<city> sorted_cities = omap.get_settings().cities;
+    int static_city_count = omap.get_settings().cities.size();
+    if( get_option<bool>( "SELECT_STARTING_CITY" ) && static_city_count > 0 ) {
+        if( u.random_start_city ) {
+            world_origin = random_entry( sorted_cities ).pos_om;
+        } else {
+            //std::vector<city> sorted_cities( sorted_cities.begin(), sorted_cities.end() );
+            const auto cities_cmp_population = []( city a, city b ) {
```
Comment (BrettDong) [inline src/game.cpp:709]:
```suggestion
            const auto cities_cmp_population = []( const city &a, const city &b ) {
```

File: src/game.cpp [hunk 3]
```diff
@@ -697,7 +697,40 @@ bool game::start_game()
...
+            uilist select_city;
+            select_city.desc_enabled = true;
+            select_city.allow_cancel = false;
+            select_city.title = _( "Select a starting city" );
+            for( const auto &c : sorted_cities ) {
+                uilist_entry entry( c.id, true, -1, c.name,
+                                    string_format(
+                                        _( "Location: <color_white>%s</color>:<color_white>%s</color>" ),
+                                        c.pos_om.to_string(), c.pos.to_string(), c.size ),
+                                    string_format( _( "(pop <color_white>%s</color>)" ), c.population ) );
```
Comment (BrettDong) [inline src/game.cpp:723]: Add a translator note:
```suggestion
                                    //~ "pop" refers to population count
                                    string_format( _( "(pop <color_white>%s</color>)" ), c.population ) );
```

File: src/game.cpp [hunk 4]
```diff
@@ -697,7 +697,40 @@ bool game::start_game()
...
+
+            uilist select_city;
+            select_city.desc_enabled = true;
+            select_city.allow_cancel = false;
+            select_city.title = _( "Select a starting city" );
+            for( const auto &c : sorted_cities ) {
+                uilist_entry entry( c.id, true, -1, c.name,
+                                    string_format(
+                                        _( "Location: <color_white>%s</color>:<color_white>%s</color>" ),
+                                        c.pos_om.to_string(), c.pos.to_string(), c.size ),
```
Comment (BrettDong) [inline src/game.cpp:722]: `c.size` is unused here?

**Comment by I-am-Erk** [comment]
Do you think you're going to get this ready for testing in 0.G or should we close it for now?

**Comment by ZhilkinSerg** [comment]
> Do you think you're going to get this ready for testing in 0.G or should we close it for now?

There are still a few things I need to do to make tests pass. This should be finished in a couple of weeks or so.

**Comment by I-am-Erk** [comment]
Ooooh

**Comment by NetSysFire** [comment]
So is it canon that the game plays in MA or is it still general New England? I am mostly asking because I am still working on the foliage stuff and I am currently using "Do enough counties in NE feature this wild plant?" and restricting that to MA only would save me some time.

**Comment by I-am-Erk** [comment]
> So is it canon that the game plays in MA or is it still general New England? I am mostly asking because I am still working on the foliage stuff and I am currently using "Do enough counties in NE feature this wild plant?" and restricting that to MA only would save me some time.

Massachusetts is the default, when we need any statistics

**Comment by I-am-Erk** [comment]
I have seen enough people playtesting this in Discord now that I am comfortable merging it.

**Pull request merged**